### PR TITLE
support QoS Depth and History via ros2 topic pub/echo.

### PR DESCRIPTION
**Pull request by fujitatomoya**
close https://github.com/ros2/ros2cli/issues/523

Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@clalancette 

could you take a look at this?

**Comment by clalancette** [comment]
> could you take a look at this?

I will take a look, but I'd prefer to first fix the depth problem as discussed in https://github.com/ros2/ros2cli/issues/523#issuecomment-641316590 in first.

File: ros2topic/ros2topic/api/__init__.py [hunk 1]
```diff
@@ -146,12 +155,19 @@ def __call__(self, prefix, parsed_args, **kwargs):
 
 def qos_profile_from_short_keys(
     preset_profile: str, reliability: str = None, durability: str = None,
+    depth: unsigned_int = 0, history: str = None,
 ) -> rclpy.qos.QoSProfile:
     """Construct a QoSProfile given the name of a preset, and optional overrides."""
     # Build a QoS profile based on user-supplied arguments
     profile = rclpy.qos.QoSPresetProfiles.get_from_short_key(preset_profile)
+    if depth > 0:
+        profile.depth = depth
```
Comment (clalancette) [inline ros2topic/ros2topic/api/__init__.py:164]: I'm not so sure about this logic.  If the user passes `--qos-depth 0`, then we'll ignore their preference.  I think we'll need to use some sentinel value (maybe -1) to correspond to "the user didn't set a depth", at which point we'd use the default for the profile that they chose.

File: ros2topic/ros2topic/api/__init__.py [hunk 2]
```diff
@@ -171,6 +187,17 @@ def add_qos_arguments_to_argument_parser(
              .format(verb, default_preset))
     default_profile = rclpy.qos.QoSPresetProfiles.get_from_short_key(
         default_preset)
+    parser.add_argument(
+        '--qos-depth', metavar='N', type=unsigned_int, default = default_profile.depth,
```
Comment (clalancette) [inline ros2topic/ros2topic/api/__init__.py:191]:
```suggestion
        '--qos-depth', metavar='N', type=unsigned_int, default=default_profile.depth,
```

File: ros2topic/ros2topic/api/__init__.py [hunk 3]
```diff
@@ -171,6 +187,17 @@ def add_qos_arguments_to_argument_parser(
...
+        help='Queue size setting to {} with '
+             '(overrides depth value of --qos-profile option, default: {})'
+             .format(verb, default_profile.depth))
```
Comment (clalancette) [inline ros2topic/ros2topic/api/__init__.py:194]: The depth here isn't necessarily a default of 0 in the case of transient_local.  Not sure how we want to communicate that, though.

File: ros2topic/ros2topic/api/__init__.py (hunk 1 above)
Comment (fujitatomoya) [reply ros2topic/ros2topic/api/__init__.py:164]: makes sense to me, fixing.

File: ros2topic/ros2topic/api/__init__.py (hunk 3 above)
Comment (fujitatomoya) [reply ros2topic/ros2topic/api/__init__.py:194]: we could check the default profile and and if the durability is transient_local AND depth==0, we can show depth=1 instead?

Comment (fujitatomoya) [reply ros2topic/ros2topic/api/__init__.py:194]: I will set -1 as default value, and not printing out.

File: ros2topic/ros2topic/api/__init__.py [hunk 4]
```diff
@@ -171,6 +187,16 @@ def add_qos_arguments_to_argument_parser(
              .format(verb, default_preset))
     default_profile = rclpy.qos.QoSPresetProfiles.get_from_short_key(
         default_preset)
+    parser.add_argument(
+        '--qos-depth', metavar='N', type=int, default=-1,
+        help='Queue size setting to {} with '
+             '(overrides depth value of --qos-profile option)')
```
Comment (clalancette) [inline ros2topic/ros2topic/api/__init__.py:193]:
```suggestion
        help='Queue size setting to {} with '
             '(overrides depth value of --qos-profile option)'
             .format(verb))
```

File: ros2topic/ros2topic/api/__init__.py [hunk 5]
```diff
@@ -146,12 +155,19 @@ def __call__(self, prefix, parsed_args, **kwargs):
...
+    if depth >= 0:
+        profile.depth = depth
+    if history:
+        profile.history = rclpy.qos.QoSHistoryPolicy.get_from_short_key(history)
     if durability:
         profile.durability = rclpy.qos.QoSDurabilityPolicy.get_from_short_key(durability)
+        if profile.durability == rclpy.qos.QoSDurabilityPolicy.TRANSIENT_LOCAL and profile.depth == 0:
+            profile.depth = 1
     if reliability:
         profile.reliability = rclpy.qos.QoSReliabilityPolicy.get_from_short_key(reliability)
```
Comment (clalancette) [inline ros2topic/ros2topic/api/__init__.py:172]: I just realized this has the same problem of overriding the users' preference.  That is, if the user says `--qos-durability transient_local --qos-depth 0`, then we'll automatically override to a depth of 1.

Arguably, that's a somewhat silly thing to do, but it is still something I think we should support.  I think you can fix this by rearranging this whole block to be something more like this:

```
    if history:
        profile.history = rclpy.qos.QoSHistoryPolicy.get_from_short_key(history)
    if durability:
        profile.durability = rclpy.qos.QoSDurabilityPolicy.get_from_short_key(durability)
    if reliability:
        profile.reliability = rclpy.qos.QoSReliabilityPolicy.get_from_short_key(reliability)
    if depth >= 0:
        profile.depth = depth
    else:
        if profile.durability == rclpy.qos.QoSDurabilityPolicy.TRANSIENT_LOCAL:
            profile.depth = 1
```

File: ros2topic/ros2topic/api/__init__.py (hunk 4 above)
Comment (fujitatomoya) [reply ros2topic/ros2topic/api/__init__.py:193]: my bad, nice catch! thanks,

File: ros2topic/ros2topic/api/__init__.py (hunk 5 above)
Comment (fujitatomoya) [reply ros2topic/ros2topic/api/__init__.py:172]: that makes sense to me, i think it is better for user. appreciate for your considering!

**Comment by fujitatomoya** [comment]
@clalancette 

sorry to disturb you once again, could you take a look at https://github.com/ros2/ros2cli/pull/528/files/11020132e4efc1ecf7d4b0f96fcf099b6678b95c..a4e73056a2fad85525f715e666fff73341a62e01 ?

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11180/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6459/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9130/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11094/)

**Comment by clalancette** [comment]
@fujitatomoya It looks like tests are failing with flake8 errors.  Can you take a look?

**Comment by fujitatomoya** [comment]
@clalancette 

i will check it and get back to you.

**Comment by fujitatomoya** [comment]
@clalancette 

fix flake8 code style, confirmed with `colcon test --packages-select ros2topic` locally.

**Comment by fujitatomoya** [comment]
looks all green, do we have a go?

**Comment by clalancette** [comment]
Here's new CI.  Assuming it goes green, I'll merge:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11200/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6477/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9162/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11121/)

**Pull request merged**